packages/admin-next/dashboard/src/components/grid/data-grid/data-grid-root/data-grid-root.tsx

```diff
@@ -0,0 +1,627 @@
+    /**
+     * If we're not dragging and not selecting or there is no anchor,
+     * then we don't want to do anything.
+     */
+    if ((!isSelecting && !isDragging) || !anchor) {
+      return
+    }
+
+    const target = e.target
+
+    /**
+     * Check if the click was on a presentation element.
+     * If so, we don't want to add it to the selection.
+     */
+    if (
+      target instanceof HTMLElement &&
+      target.querySelector("[data-role=presentation]")
+    ) {
+      return
+    }
+
+    const rowIndex = parseInt(e.currentTarget.dataset.rowIndex!)
+    const columnIndex = parseInt(e.currentTarget.dataset.columnIndex!)
+
+    /**
+     * If the target column is not the same as the anchor column,
+     * we don't want to add it to the selection.
+     */
+    if (anchor?.column !== columnIndex) {
+      return
+    }
+
+    const direction =
+      rowIndex > anchor.row ? "down" : rowIndex < anchor.row ? "up" : "none"
+
+    const last = selection[selection.length - 1] ?? anchor
+
+    /**
+     * Check if the current cell is a direct neighbour of the last cell
+     * in the selection.
+     */
+    const isNeighbour = Math.abs(rowIndex - last.row) === 1
+
+    /**
+     * If the current cell is a neighbour, we can simply update
+     * the selection based on the direction.
+     */
+    if (isNeighbour) {
+      if (isSelecting) {
+        setSelection((prev) => {
+          return prev
+            .filter((cell) => {
+              if (direction === "down") {
+                return (
+                  (cell.row <= rowIndex && cell.row >= anchor.row) ||
+                  cell.row === anchor.row
+                )
+              }
+
+              if (direction === "up") {
+                return (
+                  (cell.row >= rowIndex && cell.row <= anchor.row) ||
+                  cell.row === anchor.row
+                )
+              }
+
+              return cell.row === anchor.row
+            })
+            .concat({ row: rowIndex, column: columnIndex })
+        })
+
+        return
+      }
+
+      if (isDragging) {
+        if (anchor.row === rowIndex) {
+          return
+        }
+
+        setDragSelection((prev) => {
+          return prev
+            .filter((cell) => {
+              if (direction === "down") {
+                return (
+                  (cell.row <= rowIndex && cell.row >= anchor.row) ||
+                  cell.row === anchor.row
+                )
+              }
+
+              if (direction === "up") {
+                return (
+                  (cell.row >= rowIndex && cell.row <= anchor.row) ||
+                  cell.row === anchor.row
+                )
+              }
+
+              return cell.row === anchor.row
+            })
+            .concat({ row: rowIndex, column: columnIndex })
+        })
+
+        return
+      }
+    }
+
+    /**
+     * If the current cell is not a neighbour, we instead
```

Comment:
question: is this an optimisation to separate`isNeighbour` case? Should `selectCell()` be able to solve for that too?